runtime/src/accounts_db.rs

```diff
@@ -629,6 +629,42 @@ impl AccountsDB {
         false
     }
 
+    // Reclaim older states of rooted non-zero lamports accounts as a general
+    // AccountsDB bloat mitigation and preprocess for better zero-lamport purging.
+    fn compact_before_purge_zero_lamport_accounts(
+        &self,
+        purges: &HashMap<Pubkey, Vec<(Slot, AccountInfo)>>,
+    ) {
+        let mut accounts_index = self.accounts_index.write().unwrap();
+        let mut all_pubkeys: HashSet<Pubkey> = HashSet::<Pubkey>::default();
+        for root in accounts_index.not_compacted_roots.iter() {
```

Comment:
I think the whole purge_zero_lamport thing just needs to be checked for the delta of newly-rooted slots since last purge, inspired by this https://github.com/solana-labs/solana/pull/8405. Currently, only this new non-zero lamports account compaction is checked in that manner.

We can move to delta-purge for online processing once we finish out-bound compaction in place #8337. Would that be a good direction?

Reply:
delta purge sounds good. It's a bit challenging because of the cascading situation, an account which is not part of the delta-set could be preventing a purge, so you would have to know to go look to see if that one could be purged as well.

Reply:
Hhm. Yeah, that's right.. It looks like harder than what I originally thought.